sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala

```diff
@@ -389,14 +379,16 @@ case class TungstenAggregate(
        UnsafeRow $keyTerm = (UnsafeRow) $iterTerm.getKey();
        UnsafeRow $bufferTerm = (UnsafeRow) $iterTerm.getValue();
        $outputCode
+
+       if (shouldStop()) return;
```

Comment:
Once `shouldStop()` returns true, the caller should exit the loop (via return).

`map.free()` is called only when it had consumed all the items in the loop (without return).

Will added these to the doc string of shouldStop().